pkgs/applications/graphics/pdfcpu/default.nix

```diff
@@ -2,16 +2,36 @@
 
 buildGoModule rec {
   pname = "pdfcpu";
-  version = "0.5.0";
+  version = "0.6.0";
 
   src = fetchFromGitHub {
     owner = "pdfcpu";
     repo = pname;
     rev = "v${version}";
-    sha256 = "sha256-dEAlOKjNXL7zqlll6lqGmbopjdplDR3ewMMNu9TMsmw=";
+    hash = "sha256-410DIdsPXKMxnBL+1KX/DIPktqoB/ZxgF5GXEiqLcbw=";
+    leaveDotGit = true;
+    postFetch = ''
+      cd "$out"
+      git rev-parse HEAD > $out/COMMIT
+      date -u -d "@$(git log -1 --pretty=%ct)" "+%Y-%m-%dT%H:%M:%SZ" > $out/SOURCE_DATE_EPOCH
+      find "$out" -name .git -print0 | xargs -0 rm -rf
+    '';
   };
 
-  vendorHash = "sha256-WZsm2wiKedMP0miwnzhnSrF7Qw+jqd8dnpcehlsdMCA=";
+  vendorHash = "sha256-jVX/CFf9dd9qD3gyGVDjbfINtPLCsgdjWFix4BCpDZ0=";
+
+  ldflags = [
+    "-s"
+    "-w"
+    "-X main.version=v${version}"
+  ];
+
+  # ldflags based on metadata from git and source
+  preBuild = ''
+    ldflags+=" -X main.commit=$(cat COMMIT)"
+    ldflags+=" -X main.date=$(cat SOURCE_DATE_EPOCH)"
+  '';
```

Comment:
If that is all the purpose of the above `postFetch`, maybe we can fool the ldflags to simply:

```
ldflags = [
  "-X main.date=1970-01-01"
  "-X main.commit=${version}"
];
```

Reply:
I worry that this may lead to confusion: Imagine the maintainer getting a bug report where the commit version is the same as the program version and the date being epoch. It is something they would not expect, as from their point of view date and commit should contain useful values, i.e. from the repo from which the release was built.

Commit could differ if someone built from their own clone with rebased changes.

With somewhat litte knowledge about Nix internals what are the disadvantages of using VCS clones instead of distribution archives as sources, @doronbehar?

Reply:
As for the `main.commit` information - it is trivial to infer the commit hash from the version - and definitely upstream should be able to do that themselves. As for the date of commit, that's also infer-able from the version if you trust the distribution to not do funky stuff when distributing a package built from source. I think that these attributes are relevant for those who really build from source, and not for distributors who strictly stick to releases.

Maybe I wouldn't have minded if this wasn't a package I maintain, but I do :) - I don't like the maintenance overhead of the Git clone vs the GitHub tarball and I also prefer the `ldflags` to be cleanly set. I also had a bit experience communicating with upstream, and I'm fairly sure they won't mind too much that this is how we distribute.

Perhaps we can insert the word `Nix` to indicate that Nix built and packaged the executable, with something like:

```nix
ldflags = [
  "-X main.date=1970-01-01 (Nix)"
  "-X main.commit=${version}"
];
```

Reply:
@doronbehar I've updated this PR accordingly, unfortunately pdfcpu version 0.6.0 expects `commit` to be 8 characters and panics otherwise (see [`cmd/pdfcpu/process.go:143`](https://github.com/pdfcpu/pdfcpu/blob/master/cmd/pdfcpu/process.go#L143)); PR [`pdfcpu/pdfcpu#752`](https://github.com/pdfcpu/pdfcpu/pull/752) is a proposal to remedy this. Until that PR is merged or the issue otherwise resolved the pdfcpu package needs to provide 8 characters to  `main.commit`. Because the package's version (`0.6.0`) is only 5 characters I took the liberty of padding it with a zero width space ([U+200B](https://symbl.cc/en/200B/)).

Reply:
It just occurred to me that the need to pad `${version}` is obsolete if the patch from [pdfcpu/pdfcpu#752](https://github.com/pdfcpu/pdfcpu/pull/752) included already. Yet that may break future updates, should the patch  rejected and the affected code change.

Reply:
@doronbehar [upstream "requires"](https://github.com/pdfcpu/pdfcpu/pull/752#issuecomment-1852893068) the commit to be identical on every platform and  finds that the [proposed PR](https://github.com/pdfcpu/pdfcpu/pull/752) does not serve the intention of the project.

The options I see are:
1. This package builds pdfcpu from git setting `leaveDotGit = true`, which is much less desirable from a nixpkgs maintainers point of view
2. Introduce a `commit` variable that needs to manually be updated along with `version` and `hash` accepting breaking compatibility with [nix-update](http://github.com/Mic92/nix-update), which I find is the best compromise, given the current options
3. Ignore the upstream "requirement" and set commit to `${version}`, which is much less desirable from a pdfcpu point of view
4. Do not update pdfcpu to 0.6.0, which is much less desirable from a nixpkgs users point of view

What other options did I miss?

What is your preferred path forward?

Reply:
Hmm upstream really surprised me with how strict they want their `pdfcpu version` output to be... Now after observing the interactions with them, I think I am convinced that the original solution that was proposed in this PR is indeed the best - as it:

 - Satisfies upstream
 - Should make `nix-update` able to update this package automatically (it'll take it a bit longer but that's not too bad)
 - Will still be reproducible from a Nix perspective, unless upstream will force push again, which they probably won't.

Reply:
Glad we were able to agree on an option and move forward this updating this package to 0.6.0. I've updated this PR accordingly.